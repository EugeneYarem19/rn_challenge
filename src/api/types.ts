export interface MovieBrief {
  Title: string;
  imdbID: string;
  Poster: string;
}

export interface MovieDetailed extends MovieBrief {
  Genre: string;
  Director: string;
  Actors: string;
  Plot: string;
  Ratings: {
    Source: string;
    Value: string;
  }[];
}

export interface SearchResponse {
  Search: MovieBrief[];
  totalResults: string;
  Response: string;
}

export interface MovieResponse extends MovieDetailed {
  Response: string;
}
