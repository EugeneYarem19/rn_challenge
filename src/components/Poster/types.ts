type ImageSize = "small" | "large";

export interface IPoster {
  poster: string | undefined;
  size?: ImageSize;
  smallOnEmptyPoster?: boolean;
  smallOnError?: boolean;
  style?: Record<string, unknown> | null;
}
