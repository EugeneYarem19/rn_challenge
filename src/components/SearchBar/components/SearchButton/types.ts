export interface ISearchButton {
  onPress: () => void;
}
