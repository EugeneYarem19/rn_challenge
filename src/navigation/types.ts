export type RootStackParamList = {
  Home: undefined;
  Movie: {
    id: string;
    title: string;
  };
};
