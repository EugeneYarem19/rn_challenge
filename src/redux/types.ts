export const actionTypes = {
  SEARCH_REQUEST: "SEARCH_REQUEST",
  SEARCH_SUCCESS: "SEARCH_SUCCESS",
  SEARCH_FAILED: "SEARCH_FAILED",

  FETCH_MORE_REQUEST: "FETCH_MORE_REQUEST",
  FETCH_MORE_SUCCESS: "FETCH_MORE_SUCCESS",
  FETCH_MORE_FAILED: "FETCH_MORE_FAILED",

  FETCH_MOVIE_REQUEST: "FETCH_MOVIE_REQUEST",
  FETCH_MOVIE_SUCCESS: "FETCH_MOVIE_SUCCESS",
  FETCH_MOVIE_FAILED: "FETCH_MOVIE_FAILED",
};

export const initState = {
  currentSearchTitle: "",
  foundMovies: [],
  isFetchingMore: false,
  isFetchingMovie: false,
  isSearching: false,
  isThatsAll: false,
  nextSearchPage: 2, // can be [1-100] (from API). First page will be returned by search request, so it must be equal 2 for fetch more request
  searchErrorMessage: "",
};

export interface Rating {
  source: string;
  value: string;
}

export interface Movie {
  // optional fields will be set after user will request full info about movie
  id: string;
  poster?: string; // can be empty
  title: string;
  genre?: string;
  director?: string;
  fullPlot?: string;
  cast?: string;
  ratings?: Rating[];
}

export interface MovieDetailing {
  genre: string;
  director: string;
  fullPlot: string;
  cast: string;
  ratings: Rating[];
}

export interface State {
  currentSearchTitle: string;
  foundMovies: Movie[];
  isFetchingMore: boolean;
  isFetchingMovie: boolean;
  isSearching: boolean;
  isThatsAll: boolean; // indicates that not fetched movies are on the server on our response
  nextSearchPage: number; // next page, which will be updated only on FETCH_MORE_SUCCESS or will be set to 2 on new search
  searchErrorMessage: string; // For example, "Too many results." / "Movie not found!"
}

interface IAction {
  type: string;
}

interface IResponseFailedAction extends IAction {
  payload: {
    errorMessage: string;
  };
}

export interface SearchAction extends IAction {
  payload: {
    title: string;
  };
}

export interface SearchSuccessAction extends IAction {
  payload: {
    movies: Movie[] | [];
    isThatsAll: boolean;
  };
}

export type SearchFailedAction = IResponseFailedAction;

export type FetchMoreAction = IAction;

export type FetchMoreSuccessAction = SearchSuccessAction;

export type FetchMoreFailedAction = IResponseFailedAction;

export interface FetchMovieAction extends IAction {
  payload: {
    id: string;
  };
}

export interface FetchMovieSuccessAction extends IAction {
  payload: {
    id: string;
    detailedInfo: MovieDetailing;
  };
}

export type FetchMovieFailedAction = IAction;

export type ActionTypes =
  | FetchMoreAction
  | FetchMoreSuccessAction
  | FetchMoreFailedAction
  | FetchMovieAction
  | FetchMovieSuccessAction
  | FetchMovieFailedAction
  | SearchAction
  | SearchSuccessAction
  | SearchFailedAction;
