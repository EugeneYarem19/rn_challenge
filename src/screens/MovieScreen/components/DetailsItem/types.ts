export interface IDetailsItem {
  title: string;
  value: string;
}
