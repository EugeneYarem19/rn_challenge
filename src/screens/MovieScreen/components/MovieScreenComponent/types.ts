import { Movie, } from "@redux";

export interface IMovieScreenComponent extends Movie {
  defaultTitle: string;
  isFetchingMovie: boolean;
}
