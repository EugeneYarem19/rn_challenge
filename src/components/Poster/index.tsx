import React from "react";
import ProgressImage from "react-native-image-progress";
import { Icon, } from "react-native-material-ui";
import { View, } from "react-native";

import { LoadingIndicator, } from "../LoadingIndicator";

import { IPoster, ImageSize, } from "./types";
import { styles, } from "./styles";
import { usePoster, } from "./hooks";

export const Poster: React.FC<IPoster> = ({
  poster,
  size = ImageSize.small,
  smallOnEmptyPoster = true,
  smallOnError = true,
  style,
}) => {
  const sizeStyle = size === ImageSize.small ? styles.small : styles.large;
  const { loadError, onError, } = usePoster();

  const renderPlaceholder = (additionalSizedStyle: typeof style) => (
    <View style={[styles.placeholder, additionalSizedStyle, style,]}>
      <Icon name="wallpaper" size={50} />
    </View>
  );

  const renderPosterIcon = () => {
    if (loadError) {
      return renderPlaceholder(smallOnError ? styles.small : sizeStyle);
    } else if (!poster) {
      return renderPlaceholder(smallOnEmptyPoster ? styles.small : sizeStyle);
    } else {
      return (
        <ProgressImage
          source={{ uri: poster, }}
          indicator={LoadingIndicator}
          onError={onError}
          resizeMode="contain"
          style={[styles.poster, sizeStyle, style,]}
          threshold={0}
        />
      );
    }
  };

  return renderPosterIcon();
};
