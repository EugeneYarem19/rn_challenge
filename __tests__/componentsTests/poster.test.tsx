import React from "react";

import { ImageSize, } from "@src/components/Poster/types";
import { Poster, } from "@components";

import { mockedData, } from "@tests/__mocks__";
import { snapshotTest, } from "@tests/__utils__";

describe("Poster tests", () => {
  // tests on the default props is not changed
  test("Poster must renders correctly with set poster", () => {
    snapshotTest(<Poster poster={mockedData.movies[0].poster} />);
  });

  test("Poster must renders correctly without set poster", () => {
    snapshotTest(<Poster poster="" />);
  });

  // tests on different props
  test("Poster must renders correctly with set poster and size = 'small'", () => {
    snapshotTest(<Poster poster={mockedData.movies[0].poster} size={ImageSize.small} />);
  });

  test("Poster must renders correctly with set poster and size = 'large'", () => {
    snapshotTest(<Poster poster={mockedData.movies[0].poster} size={ImageSize.large} />);
  });

  test("Poster must renders correctly without set poster and size = 'small'", () => {
    snapshotTest(<Poster poster="" size={ImageSize.small} />);
  });

  test("Poster must renders correctly without set poster and size = 'small' and smallOnEmptyPoster = 'false'", () => {
    snapshotTest(<Poster poster="" size={ImageSize.small} smallOnEmptyPoster={false} />);
  });

  test("Poster must renders correctly without set poster and size = 'large'", () => {
    snapshotTest(<Poster poster="" size={ImageSize.large} />);
  });

  test("Poster must renders correctly without set poster and size = 'large' and smallOnEmptyPoster = 'false'", () => {
    snapshotTest(<Poster poster="" size={ImageSize.large} smallOnEmptyPoster={false} />);
  });
});
