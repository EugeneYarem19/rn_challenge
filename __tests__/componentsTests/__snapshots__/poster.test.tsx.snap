// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Poster tests Poster must renders correctly with set poster 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#131d3d",
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 8,
        "height": 180,
        "width": 122,
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLoad={[Function]}
    onLoadEnd={[Function]}
    onLoadStart={[Function]}
    onProgress={[Function]}
    resizeMode="contain"
    source={
      Object {
        "uri": "http://",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#e59f2b"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;

exports[`Poster tests Poster must renders correctly with set poster and size = 'large' 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#131d3d",
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 20,
        "height": 1027.6,
        "maxHeight": 443,
        "maxWidth": 300,
        "width": 734,
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLoad={[Function]}
    onLoadEnd={[Function]}
    onLoadStart={[Function]}
    onProgress={[Function]}
    resizeMode="contain"
    source={
      Object {
        "uri": "http://",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#e59f2b"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;

exports[`Poster tests Poster must renders correctly with set poster and size = 'small' 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#131d3d",
        "overflow": "hidden",
      },
      Object {
        "borderRadius": 8,
        "height": 180,
        "width": 122,
      },
      undefined,
    ]
  }
>
  <Image
    onError={[Function]}
    onLoad={[Function]}
    onLoadEnd={[Function]}
    onLoadStart={[Function]}
    onProgress={[Function]}
    resizeMode="contain"
    source={
      Object {
        "uri": "http://",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#e59f2b"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;

exports[`Poster tests Poster must renders correctly without set poster 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#131d3d",
        "flexDirection": "column",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 8,
        "height": 180,
        "width": 122,
      },
      undefined,
    ]
  }
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": "rgba(0, 0, 0, 0.54)",
          "fontSize": 50,
        },
        null,
        Object {
          "fontFamily": "Material Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`Poster tests Poster must renders correctly without set poster and size = 'large' 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#131d3d",
        "flexDirection": "column",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 8,
        "height": 180,
        "width": 122,
      },
      undefined,
    ]
  }
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": "rgba(0, 0, 0, 0.54)",
          "fontSize": 50,
        },
        null,
        Object {
          "fontFamily": "Material Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`Poster tests Poster must renders correctly without set poster and size = 'large' and smallOnEmptyPoster = 'false' 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#131d3d",
        "flexDirection": "column",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 20,
        "height": 1027.6,
        "maxHeight": 443,
        "maxWidth": 300,
        "width": 734,
      },
      undefined,
    ]
  }
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": "rgba(0, 0, 0, 0.54)",
          "fontSize": 50,
        },
        null,
        Object {
          "fontFamily": "Material Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`Poster tests Poster must renders correctly without set poster and size = 'small' 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#131d3d",
        "flexDirection": "column",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 8,
        "height": 180,
        "width": 122,
      },
      undefined,
    ]
  }
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": "rgba(0, 0, 0, 0.54)",
          "fontSize": 50,
        },
        null,
        Object {
          "fontFamily": "Material Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`Poster tests Poster must renders correctly without set poster and size = 'small' and smallOnEmptyPoster = 'false' 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#131d3d",
        "flexDirection": "column",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 8,
        "height": 180,
        "width": 122,
      },
      undefined,
    ]
  }
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": "rgba(0, 0, 0, 0.54)",
          "fontSize": 50,
        },
        null,
        Object {
          "fontFamily": "Material Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;
